components/scream/src/share/util/scream_kokkos_utils.hpp

```diff
@@ -117,61 +121,158 @@ struct ExeSpaceUtils<Kokkos::Cuda> {
 
 /*
  * TeamUtils contains utilities for getting concurrency info for
- * thread teams.
+ * thread teams. Don't use _TeamUtilsCommonBase directly, use
+ * TeamUtils.
  */
+
 template <typename ExeSpace = Kokkos::DefaultExecutionSpace>
-class TeamUtils
+class _TeamUtilsCommonBase
 {
-  int _team_size, _num_teams;
+ protected:
+  int _team_size, _num_teams, _max_threads, _league_size;
 
-public:
+ public:
   template <typename TeamPolicy>
-  TeamUtils(const TeamPolicy& policy) : _team_size(0)
+  _TeamUtilsCommonBase(const TeamPolicy& policy)
   {
-    const int max_threads = ExeSpace::concurrency();
+    _max_threads = ExeSpace::concurrency() / ( (!is_single_precision<Real>::value && OnGpu<ExeSpace>::value) ? 2 : 1);
     const int team_size = policy.team_size();
-    _num_teams = max_threads / team_size;
-    _team_size = max_threads / _num_teams;
+    _num_teams = _max_threads / team_size;
+    _team_size = _max_threads / _num_teams;
+    _league_size = policy.league_size();
+
+    // We will never run more teams than the policy needs
+    _num_teams = _num_teams > policy.league_size() ? _league_size : _num_teams;
```

Comment:
This is a significant change. We were providing unnecessary slots to WorkspaceManager since there is no point in providing slots beyond the leage_size.